Crash parsing HTML document

htmltest is erroring out when I run it:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x30 pc=0x1111246]

goroutine 1 [running]:
github.com/wjdp/htmltest/htmldoc.(*Document).Parse(0x0)
	/home/travis/gopath/src/github.com/wjdp/htmltest/htmldoc/document.go:47 +0x26
github.com/wjdp/htmltest/htmldoc.(*Document).IsHashValid(0x0, 0xc000364b1e, 0x15, 0x127b200)
	/home/travis/gopath/src/github.com/wjdp/htmltest/htmldoc/document.go:112 +0x2b
github.com/wjdp/htmltest/htmltest.(*HTMLTest).checkInternalHash(0xc0000c5200, 0xc000161770)
	/home/travis/gopath/src/github.com/wjdp/htmltest/htmltest/check-link.go:297 +0xad
github.com/wjdp/htmltest/htmltest.(*HTMLTest).checkInternal(0xc0000c5200, 0xc000161770)
	/home/travis/gopath/src/github.com/wjdp/htmltest/htmltest/check-link.go:271 +0xcb
github.com/wjdp/htmltest/htmltest.(*HTMLTest).checkLink(0xc0000c5200, 0xc00010c700, 0xc00012df10)
	/home/travis/gopath/src/github.com/wjdp/htmltest/htmltest/check-link.go:85 +0x3f9
github.com/wjdp/htmltest/htmltest.(*HTMLTest).testDocument(0xc0000c5200, 0xc00010c700)
	/home/travis/gopath/src/github.com/wjdp/htmltest/htmltest/htmltest.go:203 +0x193
github.com/wjdp/htmltest/htmltest.(*HTMLTest).testDocuments(0xc0000c5200)
	/home/travis/gopath/src/github.com/wjdp/htmltest/htmltest/htmltest.go:182 +0x6a
github.com/wjdp/htmltest/htmltest.Test(0xc000075aa0, 0x1, 0x1, 0x49)
	/home/travis/gopath/src/github.com/wjdp/htmltest/htmltest/htmltest.go:142 +0x8fb
main.run(0xc000075aa0, 0xc000075aa0)
	/home/travis/gopath/src/github.com/wjdp/htmltest/main.go:159 +0x1e8
main.main()
	/home/travis/gopath/src/github.com/wjdp/htmltest/main.go:66 +0x298
```

# To Reproduce
Steps to reproduce the behaviour:
1. Run with config and options …

```
htmltest --log-level 3
```

### .htmltest.yml

Please copy in your config file

```yml
DirectoryPath: "public"
EnforceHTTPS: false
CacheExpires: "6h"
CheckExternal: false
IgnoreDirectoryMissingTrailingSlash: true
IgnoreInternalEmptyHash: true
IgnoreDirs:
- Support
```

### Source files

I haven't been able to narrow it down yet -- my request is for htmltest to print the page it's processing to help narrow it down.

# Expected behaviour
print each page as it's being processed

## Versions
 - OS: MacOS 10.14.5
 - htmltest: 0.10.3

No wait -- the log ends at

```
pull-export/index.html
  DOCTYPE html []
 --- pull-export/index.html --> <nil>
testDocument on test/index.html
panic: runtime error: invalid memory address or nil pointer dereference
```

so which of these two is likely the culprit? `pull-export/index.html` or `test/index.html`?

It crashes on more files now. I've removed test/index.html since, but I still have others. My current run ends with

```
testDocument on installation/configuration/hidden-preferences/index.html
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x30 pc=0x1111246]
```

but it may not be something about that file in particular; if I set DirectoryPath to `public/installation/configuration` (it's usually set to `public`), it does not crash.

I'm seeing the same on Linux on Ubuntu Focal:

```bash
node@791983aec7ee:~/antora-base$ ./bin/htmltest -l0 public
htmltest started at 09:18:46 on public
========================================================================
0: DirectoryPath string = public
1: DirectoryIndex string = index.html
2: FilePath string = 
3: FileExtension string = .html
4: CheckDoctype bool = true
5: CheckAnchors bool = true
6: CheckLinks bool = true
7: CheckImages bool = true
8: CheckScripts bool = true
9: CheckMeta bool = true
10: CheckGeneric bool = true
11: CheckExternal bool = true
12: CheckInternal bool = true
13: CheckInternalHash bool = true
14: CheckMailto bool = true
15: CheckTel bool = true
16: CheckFavicon bool = false
17: CheckMetaRefresh bool = true
18: EnforceHTML5 bool = false
19: EnforceHTTPS bool = false
20: IgnoreURLs []interface {} = []
21: IgnoreDirs []interface {} = []
22: IgnoreInternalEmptyHash bool = false
23: IgnoreEmptyHref bool = false
24: IgnoreCanonicalBrokenLinks bool = true
25: IgnoreExternalBrokenLinks bool = false
26: IgnoreAltMissing bool = false
27: IgnoreDirectoryMissingTrailingSlash bool = false
28: IgnoreSSLVerify bool = false
29: IgnoreTagAttribute string = data-proofer-ignore
30: HTTPHeaders map[interface {}]interface {} = map[Accept:*/* Range:bytes=0-0]
31: TestFilesConcurrently bool = false
32: DocumentConcurrencyLimit int = 128
33: HTTPConcurrencyLimit int = 16
34: LogLevel int = 0
35: LogSort string = document
36: ExternalTimeout int = 15
37: StripQueryString bool = true
38: StripQueryExcludes []string = [fonts.googleapis.com]
39: EnableCache bool = true
40: EnableLog bool = true
41: OutputDir string = tmp/.htmltest
42: OutputCacheFile string = refcache.json
43: OutputLogFile string = htmltest.log
44: CacheExpires string = 336h
45: NoRun bool = false
46: VCREnable bool = false
47: Version string = 0.12.1
testDocument on Home/faq.html
Home/faq.html
  DOCTYPE html []
 --- Home/faq.html --> <nil>
  from cache --- Home/faq.html --> https://docs.tpwiki.com/Home/faq.html
  OK --- Home/faq.html --> https://docs.tpwiki.com/Home/faq.html
  from cache --- Home/faq.html --> https://docs.tpwiki.com
  OK --- Home/faq.html --> https://docs.tpwiki.com
  target does not exist --- Home/faq.html --> /oauth2/sign_out
testDocument on Home/index.html
Home/index.html
  DOCTYPE html []
 --- Home/index.html --> <nil>
  from cache --- Home/index.html --> https://docs.tpwiki.com/Home/index.html
  OK --- Home/index.html --> https://docs.tpwiki.com/Home/index.html
  from cache --- Home/index.html --> https://docs.tpwiki.com
  OK --- Home/index.html --> https://docs.tpwiki.com
  target does not exist --- Home/index.html --> /oauth2/sign_out
testDocument on SEL751_Arc_Flash_Protection_Settings/unstable/downloads/Downloads.html
SEL751_Arc_Flash_Protection_Settings/unstable/downloads/Downloads.html
  DOCTYPE html []
 --- SEL751_Arc_Flash_Protection_Settings/unstable/downloads/Downloads.html --> <nil>
  from cache --- SEL751_Arc_Flash_Protection_Settings/unstable/downloads/Downloads.html --> https://docs.tpwiki.com
  OK --- SEL751_Arc_Flash_Protection_Settings/unstable/downloads/Downloads.html --> https://docs.tpwiki.com
  target does not exist --- SEL751_Arc_Flash_Protection_Settings/unstable/downloads/Downloads.html --> /oauth2/sign_out
  from cache --- SEL751_Arc_Flash_Protection_Settings/unstable/downloads/Downloads.html --> https://gitlab.tpwiki.com/standard-designs/arc-flash-protection/SEL751_Arc_Flash_Protection_Settings/tree/master
  OK --- SEL751_Arc_Flash_Protection_Settings/unstable/downloads/Downloads.html --> https://gitlab.tpwiki.com/standard-designs/arc-flash-protection/SEL751_Arc_Flash_Protection_Settings/tree/master
  from cache --- SEL751_Arc_Flash_Protection_Settings/unstable/downloads/Downloads.html --> https://gitlab.tpwiki.com/standard-designs/arc-flash-protection/SEL751_Arc_Flash_Protection_Settings/issues
  OK --- SEL751_Arc_Flash_Protection_Settings/unstable/downloads/Downloads.html --> https://gitlab.tpwiki.com/standard-designs/arc-flash-protection/SEL751_Arc_Flash_Protection_Settings/issues
  from cache --- SEL751_Arc_Flash_Protection_Settings/unstable/downloads/Downloads.html --> https://gitlab.tpwiki.com/standard-designs/arc-flash-protection/SEL751_Arc_Flash_Protection_Settings/compare/master...master
  OK --- SEL751_Arc_Flash_Protection_Settings/unstable/downloads/Downloads.html --> https://gitlab.tpwiki.com/standard-designs/arc-flash-protection/SEL751_Arc_Flash_Protection_Settings/compare/master...master
testDocument on SEL751_Arc_Flash_Protection_Settings/unstable/setting_guide/Setting_Guide.html
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x30 pc=0x51c0d7]

goroutine 1 [running]:
github.com/wjdp/htmltest/htmldoc.(*Document).Parse(0x0)
        /home/travis/gopath/src/github.com/wjdp/htmltest/htmldoc/document.go:47 +0x37
github.com/wjdp/htmltest/htmldoc.(*Document).IsHashValid(...)
        /home/travis/gopath/src/github.com/wjdp/htmltest/htmldoc/document.go:112
github.com/wjdp/htmltest/htmltest.(*HTMLTest).checkInternalHash(0xc0000ce240, 0xc0003210b0)
        /home/travis/gopath/src/github.com/wjdp/htmltest/htmltest/check-link.go:325 +0xb0
github.com/wjdp/htmltest/htmltest.(*HTMLTest).checkInternal(0xc0000ce240, 0xc0003210b0)
        /home/travis/gopath/src/github.com/wjdp/htmltest/htmltest/check-link.go:299 +0x15d
github.com/wjdp/htmltest/htmltest.(*HTMLTest).checkLink(0xc0000ce240, 0xc0000fe480, 0xc0001ed0a0)
        /home/travis/gopath/src/github.com/wjdp/htmltest/htmltest/check-link.go:97 +0x5ec
github.com/wjdp/htmltest/htmltest.(*HTMLTest).testDocument(0xc0000ce240, 0xc0000fe480)
        /home/travis/gopath/src/github.com/wjdp/htmltest/htmltest/htmltest.go:204 +0x18c
github.com/wjdp/htmltest/htmltest.(*HTMLTest).testDocuments(0xc0000ce240)
        /home/travis/gopath/src/github.com/wjdp/htmltest/htmltest/htmltest.go:183 +0x65
github.com/wjdp/htmltest/htmltest.Test(0xc000013950, 0xc000010018, 0xc0000f9d48, 0x1)
        /home/travis/gopath/src/github.com/wjdp/htmltest/htmltest/htmltest.go:143 +0x89b
main.run(0xc000013950, 0xc000013950)
        /home/travis/gopath/src/github.com/wjdp/htmltest/main.go:159 +0x207
main.main()
        /home/travis/gopath/src/github.com/wjdp/htmltest/main.go:66 +0x268
```

My system is:

```bash
Linux 791983aec7ee 4.19.0-8-amd64 #1 SMP Debian 4.19.98-1 (2020-01-26) x86_64 x86_64 x86_64 GNU/Linux
```
running within a Docker container.

Happy to provide further information. This error is highly consistent and always occurs.

--- a/.gitignore
+++ b/.gitignore
@@ -45,3 +45,4 @@ dist/
 
 # JetBrains config
 .idea/
+*.iml

--- a/htmldoc/document.go
+++ b/htmldoc/document.go
@@ -2,11 +2,12 @@ package htmldoc
 
 import (
 	"fmt"
-	"github.com/wjdp/htmltest/output"
-	"golang.org/x/net/html"
 	"os"
 	"path"
 	"sync"
+
+	"github.com/wjdp/htmltest/output"
+	"golang.org/x/net/html"
 )
 
 // Document struct, representation of a document within the tested site
@@ -14,6 +15,7 @@ type Document struct {
 	FilePath           string                // Relative to the shell session
 	SitePath           string                // Relative to the site root
 	BasePath           string                // Base for relative links
+	IgnoreTest         bool                  // Ignore this Document for testing.
 	htmlMutex          *sync.Mutex           // Controls access to htmlNode
 	htmlNode           *html.Node            // Parsed output
 	hashMap            map[string]*html.Node // Map of valid id/names of nodes
@@ -43,6 +45,7 @@ func (doc *Document) Init() {
 // already been done. Thread safe. Either called when the document is tested
 // or when another document needs data from this one.
 func (doc *Document) Parse() {
+
 	// Only one routine may parse the doc
 	doc.htmlMutex.Lock()
 	defer doc.htmlMutex.Unlock()

--- a/htmldoc/document_store.go
+++ b/htmldoc/document_store.go
@@ -4,10 +4,11 @@
 package htmldoc
 
 import (
-	"github.com/wjdp/htmltest/output"
 	"os"
 	"path"
 	"regexp"
+
+	"github.com/wjdp/htmltest/output"
 )
 
 // DocumentStore struct, store of Documents including Document discovery
@@ -55,11 +56,6 @@ func (dS *DocumentStore) isDirIgnored(dir string) bool {
 
 // Recursive function to discover documents by walking the file tree
 func (dS *DocumentStore) discoverRecurse(dPath string) {
-	// Recurse over relative path dPath, saves found documents to dS
-	if dS.isDirIgnored(dPath) {
-		return
-	}
-
 	// Open directory to scan
 	f, err := os.Open(path.Join(dS.BasePath, dPath))
 	output.CheckErrorPanic(err)
@@ -83,9 +79,10 @@ func (dS *DocumentStore) discoverRecurse(dPath string) {
 			} else if path.Ext(fileinfo.Name()) == dS.DocumentExtension {
 				// If a file, create and save document
 				newDoc := &Document{
-					FilePath: path.Join(dS.BasePath, fPath),
-					SitePath: fPath,
-					BasePath: dPath,
+					FilePath:   path.Join(dS.BasePath, fPath),
+					SitePath:   fPath,
+					BasePath:   dPath,
+					IgnoreTest: dS.isDirIgnored(dPath),
 				}
 				newDoc.Init()
 				dS.AddDocument(newDoc)

--- a/htmldoc/document_store_test.go
+++ b/htmldoc/document_store_test.go
@@ -1,8 +1,9 @@
 package htmldoc
 
 import (
-	"github.com/daviddengcn/go-assert"
 	"testing"
+
+	"github.com/daviddengcn/go-assert"
 )
 
 func TestDocumentStoreDiscover(t *testing.T) {
@@ -25,7 +26,7 @@ func TestDocumentStoreIgnorePatterns(t *testing.T) {
 	dS.IgnorePatterns = []interface{}{"^lib/"}
 	dS.Discover()
 	// Fixtures dir has seven documents in various folders, (one ignored in lib)
-	assert.Equals(t, "document count", len(dS.Documents), 5)
+	assert.Equals(t, "document count", len(dS.Documents), 6)
 }
 
 func TestDocumentStoreDocumentExists(t *testing.T) {

--- a/htmldoc/reference.go
+++ b/htmldoc/reference.go
@@ -1,10 +1,11 @@
 package htmldoc
 
 import (
-	"golang.org/x/net/html"
 	"net/url"
 	"path"
 	"strings"
+
+	"golang.org/x/net/html"
 )
 
 // Reference struct, representation of the link between a document and a
@@ -19,6 +20,7 @@ type Reference struct {
 // NewReference : Create a new reference given a document, node and path.
 // Generates the URL object.
 func NewReference(document *Document, node *html.Node, path string) (*Reference, error) {
+
 	// Clean path
 	path = strings.TrimLeftFunc(path, invalidPrePostRune)
 	path = strings.TrimRightFunc(path, invalidPrePostRune)

--- a/htmltest/check-link.go
+++ b/htmltest/check-link.go
@@ -333,8 +333,9 @@ func (hT *HTMLTest) checkInternalHash(ref *htmldoc.Reference) {
 
 	if len(ref.URL.Path) > 0 {
 		// internal
-		refDoc, _ := hT.documentStore.ResolveRef(ref)
-		if !refDoc.IsHashValid(ref.URL.Fragment) {
+		refDoc, ok := hT.documentStore.ResolveRef(ref)
+
+		if !ok || !refDoc.IsHashValid(ref.URL.Fragment) {
 			hT.issueStore.AddIssue(issues.Issue{
 				Level:     issues.LevelError,
 				Message:   "hash does not exist",

--- a/htmltest/htmltest.go
+++ b/htmltest/htmltest.go
@@ -6,17 +6,18 @@ import (
 	"crypto/tls"
 	"errors"
 	"fmt"
-	"github.com/wjdp/htmltest/htmldoc"
-	"github.com/wjdp/htmltest/issues"
-	"github.com/wjdp/htmltest/output"
-	"github.com/wjdp/htmltest/refcache"
-	"gopkg.in/seborama/govcr.v2"
 	"net/http"
 	"os"
 	"path"
 	"strings"
 	"sync"
 	"time"
+
+	"github.com/wjdp/htmltest/htmldoc"
+	"github.com/wjdp/htmltest/issues"
+	"github.com/wjdp/htmltest/output"
+	"github.com/wjdp/htmltest/refcache"
+	"gopkg.in/seborama/govcr.v2"
 )
 
 // Base path for VCR cassettes, relative to this package
@@ -186,6 +187,14 @@ func (hT *HTMLTest) testDocuments() {
 }
 
 func (hT *HTMLTest) testDocument(document *htmldoc.Document) {
+	if document.IgnoreTest {
+		hT.issueStore.AddIssue(issues.Issue{
+			Level:   issues.LevelDebug,
+			Message: "ignored " + document.SitePath,
+		})
+		return
+	}
+
 	hT.issueStore.AddIssue(issues.Issue{
 		Level:   issues.LevelDebug,
 		Message: "testDocument on " + document.SitePath,

